src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/CompilerUtilities/ImmutableHashMapExtensions.cs

```diff
@@ -18,13 +18,14 @@ internal static class ImmutableHashMapExtensions
         /// <typeparam name="TArg">The type of argument supplied to the value factory.</typeparam>
         /// <param name="location">The variable or field to atomically update if the specified <paramref name="key" /> is not in the dictionary.</param>
         /// <param name="key">The key for the value to retrieve or add.</param>
-        /// <param name="valueFactory">The function to execute to obtain the value to insert into the dictionary if the key is not found.</param>
+        /// <param name="valueProvider">The function to execute to obtain the value to insert into the dictionary if the key is not found. Returns null if the value can't be obtained.</param>
         /// <param name="factoryArgument">The argument to pass to the value factory.</param>
-        /// <returns>The value obtained from the dictionary or <paramref name="valueFactory" /> if it was not present.</returns>
-        public static TValue GetOrAdd<TKey, TValue, TArg>(ref ImmutableHashMap<TKey, TValue> location, TKey key, Func<TKey, TArg, TValue> valueFactory, TArg factoryArgument)
+        /// <returns>The value obtained from the dictionary or <paramref name="valueProvider" /> if it was not present.</returns>
+        public static TValue? GetOrAdd<TKey, TValue, TArg>(ref ImmutableHashMap<TKey, TValue> location, TKey key, Func<TKey, TArg, TValue?> valueProvider, TArg factoryArgument)
```

Comment:
Is there any way to not have to change this to an explicit TValue? returning type? Because if you're using this and your valueProvider isn't null returning, you now are faced with the null coming back that can't happen?

Reply:
I don't think so, unless we change the return type of the method and of the valueProvider to Optional<TValue>.

Reply:
I think it's ok. It's not used that much.